server/src/main/java/org/apache/druid/segment/loading/StorageLocation.java

```diff
@@ -72,65 +83,81 @@ public long getMaxSize()
+   * Remove a segment file from this location. The given file argument must be a file rather than directory.
    */
-  public synchronized void addFile(File file)
+  public synchronized void removeFile(File file)
   {
-    if (file.isDirectory()) {
-      throw new ISE("[%s] must be a file. Use a");
-    }
-    if (files.add(file)) {
-      currSize += FileUtils.sizeOf(file);
+    if (files.remove(file)) {
+      currSize -= FileUtils.sizeOf(file);
+    } else {
+      log.warn("File[%s] is not found under this location[%s]", file, path);
     }
   }
 
   /**
-   * Add a new segment dir to this location. The segment size is added to currSize.
+   * Remove a segment dir from this location. The segment size is subtracted from currSize.
    */
-  public synchronized void addSegmentDir(File segmentDir, DataSegment segment)
+  public synchronized void removeSegmentDir(File segmentDir, DataSegment segment)
   {
-    if (files.add(segmentDir)) {
-      currSize += segment.getSize();
+    if (files.remove(segmentDir)) {
+      currSize -= segment.getSize();
+    } else {
+      log.warn("SegmentDir[%s] is not found under this location[%s]", segmentDir, path);
     }
   }
 
   /**
-   * Remove a segment file from this location. The given file argument must be a file rather than directory.
+   * Reserves space to store the given segment. The segment size is added to currSize.
+   * If it succeeds, it returns a file for the given segmentDir in this storage location. Returns null otherwise.
    */
-  public synchronized void removeFile(File file)
+  @Nullable
+  public synchronized File reserve(String segmentDir, DataSegment segment)
   {
-    if (files.remove(file)) {
-      currSize -= FileUtils.sizeOf(file);
-    }
+    return reserve(segmentDir, segment.getId(), segment.getSize());
   }
 
   /**
-   * Remove a segment dir from this location. The segment size is subtracted from currSize.
+   * Reserves space to store the given segment.
+   * If it succeeds, it returns a file for the given segmentFilePathToAdd in this storage location.
+   * Returns null otherwise.
    */
-  public synchronized void removeSegmentDir(File segmentDir, DataSegment segment)
+  @Nullable
+  public synchronized File reserve(String segmentFilePathToAdd, SegmentId segmentId, long segmentSize)
   {
-    if (files.remove(segmentDir)) {
-      currSize -= segment.getSize();
+    final File segmentFileToAdd = new File(path, segmentFilePathToAdd);
+    if (files.contains(segmentFileToAdd)) {
+      return null;
+    }
+    if (canHandle(segmentId, segmentSize)) {
+      files.add(segmentFileToAdd);
+      currSize += segmentSize;
+      return segmentFileToAdd;
+    } else {
+      return null;
     }
   }
 
-  public boolean canHandle(DataSegment segment)
+  /**
+   * This method is available for only unit tests. Production code must use {@link #reserve} instead.
+   */
+  @VisibleForTesting
+  boolean canHandle(SegmentId segmentId, long segmentSize)
```

Comment:
This method should be annotated `@GuardedBy("this")`. See https://github.com/code-review-checklists/java-concurrency#guarded-by

Reply:
Added.